_posts/2019-04-23-autoregressive-flows/experimenting-with-autoregressive-flows.Rmd

```diff
@@ -0,0 +1,645 @@
+
+```{r eval=TRUE, echo=FALSE, fig.cap="Figure from Masked Autoregressive Flow for Density Estimation[@2017arXiv170507057P]"}
+knitr::include_graphics("images/maf_results.png")
+```
+
+
+Collecting information from the paper, we know that
+
+- data was included from the file _ethylene_CO.txt_ only;
+- discrete columns were eliminated, as well as all columns with correlations > .98; and
+- the remaining 8 columns ^[not specified individually in the paper] were standardised (z-transformed).
+
+Regarding the neural network architecture, we gather that
+
+- each of the 10 MAF layers was followed by a batchnorm;
+- as to feature order, the first MAF layer used the variable order that came with the dataset; then every consecutive layer reversed it;
+- specifically for this dataset and as opposed to all other UCI datasets, _tanh_ was used for activation instead of _relu_;
+- the Adam optimizer was used, with a learning rate of 1e-4;
+- there were two hidden layers for each MAF, with 100 units each;
+- training went on until no improvement occurred for 30 consecutive epochs on the validation set; and 
+- the base distribution was a multivariate Gaussian.
+
+This is all useful information for our attempt to estimate this dataset, but the essential bit is this. In case you knew the dataset already, you might have been wondering how the authors would deal with the dimensionality of the data: It is a time series, and the MADE architecture explored above introduces autoregressivity between features, not time steps. So how is the additional temporal autoregressivity to be handled? The answer is: The time dimension is essentially removed. In the authors' words, 
+
+> [...] it is a time series but was treated as if each example were an i.i.d. sample from the marginal distribution. 
+
+This undoubtedly is useful information for our present modeling attempt, but it also tells us something else: We might have to look beyond MADE layers for actual time series modeling.
+
+Now though let's look at this example of using MAF for multivariate modeling, with no time or spatial dimension to be taken into account.
+
+Following the hints the authors gave us, this is what we do.
+
+```{r}
+# load libraries -------------------------------------------------------------
+library(tensorflow)
+library(tfprobability)
+
+tfe_enable_eager_execution(device_policy = "silent")
+
+library(tfdatasets)
+library(dplyr)
+library(readr)
+library(purrr)
+library(caret)
+
+# read data ------------------------------------------------------------------
+df <- read_table2("ethylene_CO.txt",
+                  skip = 1,
+                  col_names = FALSE)
+
+# we don't know if we'll end up with the same columns as the authors did,
+# but we try (at least we do end up with 8 columns)
+df <- df[,-(1:3)]
+hc <- findCorrelation(cor(df), cutoff = 0.985)
+df2 <- df[,-c(hc)]
+
+# scale
+df2 <- scale(df2)
```

Comment:
It may be helpful to see some summaries or a `glimpse()` of the data so one can follow more easily.

Reply:
right, adding...